Lib/asyncio/base_events.py

```diff
@@ -647,6 +651,76 @@ def _getaddrinfo_debug(self, host, port, family, type, proto, flags):
         return await self.run_in_executor(
             None, socket.getnameinfo, sockaddr, flags)
 
+    async def sock_sendfile(self, sock, file, offset=0, count=None,
+                            *, fallback=True):
+        if self._debug and sock.gettimeout() != 0:
+            raise ValueError("the socket must be non-blocking")
+        self._check_sendfile_params(sock, file, offset, count)
+        if fallback:
```

Comment:
The logic must be different.

Currently you use slow fallback code if `fallback=True`.

Instead you should always try to use fast syscall first, and if it fails, we check if the fallback is enabled.  If it is, we emulate sendfile, if not -- we raise an error.